tests/src/mender-update/cli/cli_test.cpp

```diff
@@ -2174,8 +2174,8 @@ exit 0
 			"--datastore",
 			tmpdir.Path(),
 			"install",
-			"--stop-after",
-			"ArtifactCommit",
+			"--stop-before",
+			"ArtifactCommit_Leave",
```

Comment:
What is the difference between stopping before `ArtifactCommit_Leave` or `Cleanup`? For the successful case, there is no real difference. And yes, for failure it will behave different (it would stop or not), but maybe we could merge these two use cases and stop only before `Cleanup`?

From the API point of view, to allow stopping before a _Leave_ state sounds strange. Note that every other allowed state to stop before are _Enter_ states (or Cleanup)

Reply:
There is a big difference actually, but I also didn't realize it until I started looking at this in great detail. `ArtifactCommit_Leave` state scripts are only executed when the commit has definitely happened, and will not roll back (this is documented [here](https://docs.mender.io/3.7/artifact-creation/state-scripts#power-loss)). In other words, when executing `ArtifactCommit_Leave` scripts, we need to be sure that we will not roll back afterwards, hence we need to stop before them, and not before `Cleanup`, at which point it would be too late already. Rolling back after a successful commit, but before the Leave scripts, is possible with the orchestrator because *other* payloads may fail to commit in between.

The state machine is indeed quite complicated in this respect, and I can't say I particularly like this complexity. But it underlines how hard it is to both execute all states correctly, including on powerloss, while simultaneously offering a command line interface to fine control those states. It's not easy! 😃

Reply:
Okey. Got it :memo: